### ISSUE TITLE
nginx - From inside of a Docker container, how do I connect to the localhost of the machine? - Stack Overflow

### ISSUE THREAD

**Issue by occoder** [occoder/common_sql#432]
**Edit:**

If you are using [Docker-for-mac](https://docs.docker.com/docker-for-mac/networking/#there-is-no-docker0-bridge-on-macos#i-want-to-connect-from-a-container-to-a-service-on-the-host) or [Docker-for-Windows](https://docs.docker.com/docker-for-windows/networking/#there-is-no-docker0-bridge-on-windows#i-want-to-connect-from-a-container-to-a-service-on-the-host) 18.03+, just connect to your mysql service using the host `host.docker.internal` (instead of the `127.0.0.1` in your connection string).

If you are using Docker-for-Linux 20.10.0+, you can also use the host `host.docker.internal` **if** you started your Docker container with the `--add-host host.docker.internal:host-gateway` option.

Otherwise, read below

* * *

## TLDR

Use `--network="host"` in your `docker run` command, then `127.0.0.1` in your docker container will point to your docker host.

Note: This mode only works on Docker for Linux, [per the documentation](https://docs.docker.com/network/host/).

* * *

Docker offers [different networking modes](https://docs.docker.com/engine/reference/run/#network-settings) when running containers. Depending on the mode you choose you would connect to your MySQL database running on the docker host differently.

## docker run --network="bridge" (default)

Docker creates a bridge named `docker0` by default. Both the docker host and the docker containers have an IP address on that bridge.

on the Docker host, type `sudo ip addr show docker0` you will have an output looking like:

```
[vagrant@docker:~] $ sudo ip addr show docker0
4: docker0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 56:84:7a:fe:97:99 brd ff:ff:ff:ff:ff:ff
    inet 172.17.42.1/16 scope global docker0
       valid_lft forever preferred_lft forever
    inet6 fe80::5484:7aff:fefe:9799/64 scope link
       valid_lft forever preferred_lft forever

```

So here my docker host has the IP address `172.17.42.1` on the `docker0` network interface.

Now start a new container and get a shell on it: `docker run --rm -it ubuntu:trusty bash` and within the container type `ip addr show eth0` to discover how its main network interface is set up:

```
root@e77f6a1b3740:/# ip addr show eth0
863: eth0: <BROADCAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 66:32:13:f0:f1:e3 brd ff:ff:ff:ff:ff:ff
    inet 172.17.1.192/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::6432:13ff:fef0:f1e3/64 scope link
       valid_lft forever preferred_lft forever

```

Here my container has the IP address `172.17.1.192`. Now look at the routing table:

```
root@e77f6a1b3740:/# route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         172.17.42.1     0.0.0.0         UG    0      0        0 eth0
172.17.0.0      *               255.255.0.0     U     0      0        0 eth0

```

So the IP Address of the docker host `172.17.42.1` is set as the default route and is accessible from your container.

```
root@e77f6a1b3740:/# ping 172.17.42.1
PING 172.17.42.1 (172.17.42.1) 56(84) bytes of data.
64 bytes from 172.17.42.1: icmp_seq=1 ttl=64 time=0.070 ms
64 bytes from 172.17.42.1: icmp_seq=2 ttl=64 time=0.201 ms
64 bytes from 172.17.42.1: icmp_seq=3 ttl=64 time=0.116 ms

```

## docker run --network="host"

Alternatively you can run a docker container with [network settings set to `host`](http://docs.docker.com/engine/reference/run/#network-host). Such a container will share the network stack with the docker host and from the container point of view, `localhost` (or `127.0.0.1`) will refer to the docker host.

Be aware that any port opened in your docker container would be opened on the docker host. And this without requiring the [`-p` or `-P` `docker run` option](https://docs.docker.com/engine/reference/run/#expose-incoming-ports).

IP config on my docker host:

```
[vagrant@docker:~] $ ip addr show eth0
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:98:dc:aa brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fe98:dcaa/64 scope link
       valid_lft forever preferred_lft forever

```

and from a docker container in **host** mode:

```
[vagrant@docker:~] $ docker run --rm -it --network=host ubuntu:trusty ip addr show eth0
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:98:dc:aa brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fe98:dcaa/64 scope link
       valid_lft forever preferred_lft forever

```

As you can see both the docker host and docker container share the exact same network interface and as such have the same IP address.

* * *

## bridge mode

To access MySQL running on the docker host from containers in _bridge mode_, you need to make sure the MySQL service is listening for connections on the `172.17.42.1` IP address.

To do so, make sure you have either `bind-address = 172.17.42.1` or `bind-address = 0.0.0.0` in your MySQL config file (my.cnf).

If you need to set an environment variable with the IP address of the gateway, you can run the following code in a container :

```
export DOCKER_HOST_IP=$(route -n | awk '/UG[ \t]/{print $2}')

```

then in your application, use the `DOCKER_HOST_IP` environment variable to open the connection to MySQL.

**Note:** if you use `bind-address = 0.0.0.0` your MySQL server will listen for connections on all network interfaces. That means your MySQL server could be reached from the Internet ; make sure to setup firewall rules accordingly.

**Note 2:** if you use `bind-address = 172.17.42.1` your MySQL server won't listen for connections made to `127.0.0.1`. Processes running on the docker host that would want to connect to MySQL would have to use the `172.17.42.1` IP address.

## host mode

To access MySQL running on the docker host from containers in _host mode_, you can keep `bind-address = 127.0.0.1` in your MySQL configuration and all you need to do is to connect to `127.0.0.1` from your containers:

```
[vagrant@docker:~] $ docker run --rm -it --network=host mysql mysql -h 127.0.0.1 -uroot -p
Enter password:
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 36
Server version: 5.5.41-0ubuntu0.14.04.1 (Ubuntu)

Copyright (c) 2000, 2014, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its affiliates. Other names may be trademarks of their respective owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql>

```

**note:** Do use `mysql -h 127.0.0.1` and not `mysql -h localhost`; otherwise the MySQL client would try to connect using a unix socket. 
 [https://stackoverflow.com/questions/24319662/from-inside-of-a-docker-container-how-do-i-connect-to-the-localhost-of-the-mach/57833573#57833573](https://stackoverflow.com/questions/24319662/from-inside-of-a-docker-container-how-do-i-connect-to-the-localhost-of-the-mach/57833573#57833573)
